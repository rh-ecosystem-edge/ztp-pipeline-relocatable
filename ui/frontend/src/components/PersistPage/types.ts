export type PersistErrorType = {
  title: string;
  message: string;
} | null;
