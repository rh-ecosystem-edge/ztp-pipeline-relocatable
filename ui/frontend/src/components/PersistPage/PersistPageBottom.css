.persist-page-bottom {
  align-items: center;
}

.persist-page-bottom__success-icon {
  height: var(--pf-global--icon--FontSize--xl);
  width: var(--pf-global--icon--FontSize--xl);
}

.persist-page-bottom__progress-icon {
  height: var(--pf-global--icon--FontSize--xl);
  width: var(--pf-global--icon--FontSize--xl);
}

.persist-page-bottom__error-item {
  width: 40rem;
}

.persist-page-bottom__error-item-footer {
  height: 81px; /* TODO: make it responsive! */

  display: block;
  margin-left: auto;
  margin-right: auto;
}

.persist-page-bottom__persist-progress {
  width: 40rem;
}
