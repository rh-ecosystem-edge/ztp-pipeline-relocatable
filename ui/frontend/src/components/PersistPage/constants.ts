export const RESOURCE_CREATE_TITLE = 'Resource create failed';
export const RESOURCE_PATCH_TITLE = 'Resource update failed';
export const MISSING_VALUE = 'Missing value';
export const RESOURCE_FETCH_TITLE = 'Failed to read resource';
export const PERSIST_IDP = 'Registering new identity HTPasswd provider failed.';
export const KUBEADMIN_REMOVE = 'Kubeadmin removal failed.';
export const KUBEADMIN_REMOVE_OK = 'Kubeadmin removed.';
export const PERSIST_DOMAIN = 'Changing domain failed.';
export const UI_POD_NOT_READY = 'Configuration UI pod is not ready';
export const API_LIVENESS_FAILED_TITLE = 'API can not be reached';
export const WAIT_ON_OPERATOR_TITLE = 'Reading operator status failed';

export const IDENTITY_PROVIDER_NAME = 'ztpfw-htpasswd-idp';
export const ZTPFW_NAMESPACE = 'ztpfw-ui';

export const DELAY_BEFORE_FINAL_REDIRECT = 10 * 1000;
export const DELAY_BEFORE_QUERY_RETRY = 5 * 1000; /* ms */
export const MAX_LIVENESS_CHECK_COUNT = 20 * ((60 * 1000) / DELAY_BEFORE_QUERY_RETRY); // max 20 minutes

export const SSH_PRIVATE_KEY_SECRET = {
  name: 'cluster-ssh-keypair',
  namespace: 'default' /* !?! */,
};
export const SSH_PRIVATE_KEY_SECRET_TITLE = 'Missing SSH private key';
export const SSH_PRIVATE_KEY_SECRET_INCORRECT = 'Incorrect SSH key secret';

export const ADDRESS_POOL_ANNOTATION_KEY = 'metallb.universe.tf/address-pool';
export const ADDRESS_POOL_NAMESPACE = 'metallb';

export const kubeadminSecret = { name: 'kubeadmin', namespace: 'kube-system' };
