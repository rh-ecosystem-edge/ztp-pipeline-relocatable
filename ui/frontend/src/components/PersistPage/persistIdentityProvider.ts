import { cloneDeep } from 'lodash';
import { createResource, patchResource, postRequest } from '../../resources';
import {
  getHtpasswdIdentityProvider,
  getOAuth,
  IndetityProviderType,
  OAuth,
} from '../../resources/oauth';
import { deleteSecret, getSecret } from '../../resources/secret';
import { IResource, Secret, PatchType } from '../../backend-shared';
import {
  IDENTITY_PROVIDER_NAME,
  kubeadminSecret,
  KUBEADMIN_REMOVE,
  PERSIST_IDP,
  RESOURCE_CREATE_TITLE,
  RESOURCE_FETCH_TITLE,
  RESOURCE_PATCH_TITLE,
} from './constants';
import { CLUSTER_ADMIN_ROLE_BINDING, HTPASSWD_SECRET } from './resourceTemplates';
import { PersistErrorType } from './types';
import { PersistSteps, UsePersistProgressType } from '../PersistProgress';

const getHtpasswdData = async (
  setError: (error: PersistErrorType) => void,
  username: string,
  password: string,
): Promise<string | undefined> => {
  try {
    const htPasswdData = (await postRequest('/htpasswd', {
      username,
      password,
    }).promise) as { htpasswdData: string };
    if (!htPasswdData?.htpasswdData) {
      console.error('Can not encode password to htpasswd');
      setError({
        title: RESOURCE_CREATE_TITLE,
        message: `Can not encode password for the ${HTPASSWD_SECRET.metadata.name} htpasswd secret in the ${HTPASSWD_SECRET.metadata.name} namespace.`,
      });
      return undefined;
    }
    const htPasswdDataB64 = btoa(htPasswdData.htpasswdData);
    return htPasswdDataB64;
  } catch (e) {
    console.error(e);
    setError({
      title: PERSIST_IDP,
      message: 'Failed to add new OAuth identity provider.',
    });
  }

  return undefined;
};

const createSecret = async (
  setError: (error: PersistErrorType) => void,
  htPasswdDataB64: string,
): Promise<Secret | undefined> => {
  try {
    const object = cloneDeep(HTPASSWD_SECRET);
    object.data && (object.data.htpasswd = htPasswdDataB64);

    return createResource(object).promise;
  } catch (e) {
    console.error('Can not create resource: ', e);
    setError({
      title: RESOURCE_CREATE_TITLE,
      message: `Can not create ${HTPASSWD_SECRET.metadata.generateName} htpasswd secret in the ${HTPASSWD_SECRET.metadata.namespace} namespace.`,
    });
  }
  return undefined;
};

const patchIDP = async (
  setError: (error: PersistErrorType) => void,
  oauth: OAuth,
  secret: Secret,
): Promise<OAuth | undefined> => {
  const identityProviders: IndetityProviderType[] = oauth.spec?.identityProviders || [];
  identityProviders.push({
    name: IDENTITY_PROVIDER_NAME,
    mappingMethod: 'claim',
    type: 'HTPasswd',
    htpasswd: {
      fileData: {
        name: secret.metadata.name || '',
      },
    },
  });
  const patches: PatchType[] = [
    {
      op: oauth.spec?.identityProviders === undefined ? 'add' : 'replace',
      path: '/spec/identityProviders',
      value: identityProviders,
    },
  ];

  try {
    const response = await patchResource(oauth, patches).promise;
    console.log(`Patched OAuth resource for new htpasswd provider:`, response);
    return response;
  } catch (e) {
    console.error('Can not patch resource: ', e, oauth, patches);
    setError({
      title: RESOURCE_PATCH_TITLE,
      message: `Can not update OAuth resource for new htpasswd identity provider.`,
    });
  }
  return undefined;
};

// Mimics: oc adm policy add-cluster-role-to-user cluster-admin [USER]
const bindClusterAdminRole = async (
  setError: (error: PersistErrorType) => void,
  username: string,
): Promise<IResource | undefined> => {
  try {
    const object = cloneDeep(CLUSTER_ADMIN_ROLE_BINDING);
    object.subjects[0].name = username;

    return createResource(object).promise;
  } catch (e) {
    console.error('Can not create resource: ', e);
    setError({
      title: RESOURCE_CREATE_TITLE,
      message: `Can not create ${CLUSTER_ADMIN_ROLE_BINDING.metadata.generateName} ClusterRoleBinding resource to grant cluster-admin role to the ${username} user.`,
    });
  }
  return undefined;
};

export enum PersistIdentityProviderResult {
  error = 'error',
  skipped = 'skipped',
  userCreated = 'userCreated',
}

export const persistIdentityProvider = async (
  setError: (error: PersistErrorType) => void,
  setProgress: UsePersistProgressType['setProgress'],
  username: string,
  password: string,
): Promise<PersistIdentityProviderResult> => {
  if (!username || !password) {
    console.log('persistIdentityProvider: username or password missing, so skipping that step.');
    setProgress(PersistSteps.PersistIDP);
    return PersistIdentityProviderResult.skipped;
  }

  let oauth;
  try {
    // cluster-scoped, name: cluster
    oauth = await getOAuth().promise;
  } catch (e) {
    console.error(e);
    setError({ title: RESOURCE_FETCH_TITLE, message: 'Failed to get the OAuth resource.' });
    return PersistIdentityProviderResult.error;
  }

  const htpasswdIdentityProvider = getHtpasswdIdentityProvider(oauth);

  if (htpasswdIdentityProvider) {
    console.info(
      `The ${IDENTITY_PROVIDER_NAME} is already present, skipping. To edit username/password, change the htpasswd secret resource.`,
    );

    // skip username/passwod wizard steps for that case
    setProgress(PersistSteps.PersistIDP);
    return PersistIdentityProviderResult.skipped;
  }

  // encode password
  const htPasswdDataB64 = await getHtpasswdData(setError, username, password);
  if (!htPasswdDataB64) {
    return PersistIdentityProviderResult.error;
  }

  // create HTPasswd Secret
  const secret = await createSecret(setError, htPasswdDataB64);
  if (!secret) {
    return PersistIdentityProviderResult.error;
  }

  // Patch (add) new IDP record
  if (!(await patchIDP(setError, oauth, secret))) {
    return PersistIdentityProviderResult.error;
  }

  // grant cluster-admin privileges
  if (!(await bindClusterAdminRole(setError, username))) {
    return PersistIdentityProviderResult.error;
  }

  setProgress(PersistSteps.PersistIDP);
  return PersistIdentityProviderResult.userCreated;
};

export const deleteKubeAdmin = async (
  setError: (error: PersistErrorType) => void,
): Promise<boolean> => {
  let secret;
  try {
    // The 404 is a valid state in this flow
    secret = await getSecret(kubeadminSecret).promise;

    // Ok, it is still there, so remove it
    await deleteSecret(kubeadminSecret).promise;
  } catch (e) {
    if (secret) {
      // It is not a failure if the secret is already missing
      console.error(e);
      setError({ title: KUBEADMIN_REMOVE, message: 'Failed to remove the kubeadmin user.' });
      return false;
    }
  }

  return true;
};
