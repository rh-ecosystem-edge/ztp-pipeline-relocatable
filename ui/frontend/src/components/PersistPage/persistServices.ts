import { cloneDeep } from 'lodash';

import { createResource, patchResource } from '../../resources';
import { PatchType, Service } from '../../backend-shared';
import { getService } from '../../resources/service';
import { addIpDots } from '../utils';
import {
  ADDRESS_POOL_ANNOTATION_KEY,
  ADDRESS_POOL_NAMESPACE,
  API_LIVENESS_FAILED_TITLE,
  MISSING_VALUE,
  RESOURCE_CREATE_TITLE,
  RESOURCE_PATCH_TITLE,
} from './constants';
import {
  ADDRESS_POOL_TEMPLATE,
  SERVICE_TEMPLATE_API,
  SERVICE_TEMPLATE_METALLB_INGRESS,
} from './resourceTemplates';
import { PersistErrorType } from './types';
import { waitForLivenessProbe } from './utils';

const createAddressPool = async (
  setError: (error: PersistErrorType) => void,
  type: 'api' | 'ingress',
  serviceIp: string,
): Promise<string | undefined> => {
  try {
    const object = cloneDeep(ADDRESS_POOL_TEMPLATE);
    object.metadata.generateName = `ztpfw-${type}-`;
    // object.metadata.namespace = namespace;
    object.spec.addresses = [`${serviceIp}-${serviceIp}`];

    const response = await createResource(object).promise;
    console.log('Resource created: ', response);

    return response.metadata.name;
  } catch (e) {
    console.error('Can not create resource: ', e);
    setError({
      title: RESOURCE_CREATE_TITLE,
      message: `Can not create ${type} AddressPool in the ${ADDRESS_POOL_NAMESPACE} namespace.`,
    });
  }
  return undefined;
};

const patchAddressPool = async (
  setError: (error: PersistErrorType) => void,
  addressPoolName: string,
  serviceIp: string,
): Promise<boolean> => {
  const addrPoolObj = {
    // no need to fetch the resource, patch it right away
    apiVersion: ADDRESS_POOL_TEMPLATE.apiVersion,
    kind: ADDRESS_POOL_TEMPLATE.kind,
    metadata: {
      name: addressPoolName,
      namespace: ADDRESS_POOL_NAMESPACE,
    },
  };
  const addrPoolPatches = [
    {
      op: 'replace',
      path: '/spec/addresses',
      value: [`${serviceIp}-${serviceIp}`],
    },
  ];

  try {
    await patchResource(addrPoolObj, addrPoolPatches).promise;
    console.log(
      `Patched ${addressPoolName} AddressPool name in the ${ADDRESS_POOL_NAMESPACE} namespace`,
    );
  } catch (e) {
    console.error('Can not patch resource: ', e, addrPoolObj, addrPoolPatches);
    setError({
      title: RESOURCE_PATCH_TITLE,
      message: `Can not update ${addressPoolName} AddressPool in the ${ADDRESS_POOL_NAMESPACE} namespace.`,
    });
    return false;
  }

  return true;
};

const saveService = async (
  setError: (error: PersistErrorType) => void,
  _serviceIp: string,
  template: Service,
  actionName: string,
  type: 'api' | 'ingress',
): Promise<boolean> => {
  if (!(_serviceIp?.replaceAll(' ', '').length >= 4)) {
    setError({
      title: MISSING_VALUE,
      message: `Missing value for ${actionName}`,
    });
    return false;
  }

  let object: Service;
  const name = template.metadata.name || '';
  const namespace = template.metadata.namespace || '';

  const serviceIp = addIpDots(_serviceIp);

  try {
    // Try to find existing service resource. To decide about Add/Patch
    object = await getService({
      name,
      namespace,
    }).promise;
  } catch (e) {
    // Service resource not found - create new one
    const addressPoolName = await createAddressPool(setError, type, serviceIp);
    if (!addressPoolName) {
      return false;
    }

    object = cloneDeep(template);
    object.spec && (object.spec.loadBalancerIP = serviceIp);
    object.metadata.annotations = object.metadata.annotations || {};
    object.metadata.annotations[ADDRESS_POOL_ANNOTATION_KEY] = addressPoolName;

    try {
      const response = await createResource(object).promise;
      console.log('Resource created: ', response);
    } catch (e) {
      console.error('Can not create resource: ', e);
      setError({
        title: RESOURCE_CREATE_TITLE,
        message: `Can not create ${name} service in the ${namespace} namespace for ${actionName}.`,
      });
      return false;
    }

    // New Service and AddressPool created
    return true;
  }

  // Patch existing Service
  if (object.spec?.loadBalancerIP !== serviceIp) {
    // if there's a change
    const patches: PatchType[] = [
      {
        op: object.spec?.loadBalancerIP === undefined ? 'add' : 'replace',
        path: '/spec/loadBalancerIP',
        value: serviceIp,
      },
    ];

    const serviceAnnotations = object.metadata.annotations || {};
    let addressPoolName: string | undefined = serviceAnnotations?.[ADDRESS_POOL_ANNOTATION_KEY];
    if (addressPoolName) {
      // The AddressPool resource already exists, patch it
      if (!(await patchAddressPool(setError, addressPoolName, serviceIp))) {
        return false;
      }
    } else {
      // The AddressPool resource does not exist yet, create it
      addressPoolName = await createAddressPool(setError, type, serviceIp);
      if (!addressPoolName) {
        return false;
      }

      serviceAnnotations[ADDRESS_POOL_ANNOTATION_KEY] = addressPoolName;
      patches.push({
        op: object.metadata.annotations === undefined ? 'add' : 'replace',
        path: `/metadata/annotations`,
        value: serviceAnnotations,
      });
    }

    try {
      const response = await patchResource(object, patches).promise;
      console.log(`Patched ${actionName} service: `, response);
    } catch (e) {
      console.error('Can not patch resource: ', e, object, patches);
      setError({
        title: RESOURCE_PATCH_TITLE,
        message: `Can not update ${name} service in the ${namespace} namespace for ${actionName}.`,
      });
      return false;
    }
  } else {
    console.log(`No changes to ${name} service detected. Skipping ${actionName}.`);
  }

  return true;
};

export const saveIngress = async (
  setError: (error: PersistErrorType) => void,
  ingressIp: string,
): Promise<boolean> =>
  saveService(setError, ingressIp, SERVICE_TEMPLATE_METALLB_INGRESS, 'ingress IP', 'ingress');

export const saveApi = async (
  setError: (error: PersistErrorType) => void,
  apiip: string,
): Promise<boolean> => {
  if (!(await saveService(setError, apiip, SERVICE_TEMPLATE_API, 'API IP', 'api'))) {
    return false;
  }

  if (!(await waitForLivenessProbe())) {
    setError({
      title: API_LIVENESS_FAILED_TITLE,
      message: 'Can not reach API on time.',
    });
    return false;
  }

  return true;
};
