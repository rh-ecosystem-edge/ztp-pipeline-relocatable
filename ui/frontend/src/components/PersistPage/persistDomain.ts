import { postRequest } from '../../resources';
import { PersistSteps, UsePersistProgressType } from '../PersistProgress';
import { PERSIST_DOMAIN } from './constants';
import { PersistErrorType } from './types';

export const persistDomain = async (
  setError: (error: PersistErrorType) => void,
  setProgress: UsePersistProgressType['setProgress'],
  domain?: string,
): Promise<boolean> => {
  if (!domain) {
    console.info('Domain change not requested, so skipping that step.');
    setProgress(PersistSteps.PersistDomain);
    return true; // skip
  }

  try {
    // Due to complexity, the flow has been moved to backend to decrease risks related to network communication
    await postRequest('/changeDomain', {
      domain,
    }).promise;
  } catch (e) {
    console.error(e);
    setError({
      title: PERSIST_DOMAIN,
      message: `Failed to change the cluster domain to "${domain}".`,
    });
    return false;
  }

  setProgress(PersistSteps.PersistDomain);
  return true;
};
