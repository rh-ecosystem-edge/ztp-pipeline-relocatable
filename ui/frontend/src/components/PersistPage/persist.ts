import { getCondition, ZTPFW_UI_ROUTE_PREFIX } from '../../copy-backend-common';
import { getClusterOperator } from '../../resources/clusteroperator';
import { PersistSteps, UsePersistProgressType } from '../PersistProgress';
import { K8SStateContextData } from '../types';
import { delay } from '../utils';
import {
  DELAY_BEFORE_FINAL_REDIRECT,
  DELAY_BEFORE_QUERY_RETRY,
  MAX_LIVENESS_CHECK_COUNT,
  UI_POD_NOT_READY,
  WAIT_ON_OPERATOR_TITLE,
} from './constants';
import { persistDomain } from './persistDomain';
import { persistIdentityProvider, PersistIdentityProviderResult } from './persistIdentityProvider';
import { saveApi, saveIngress } from './persistServices';
import { PersistErrorType } from './types';
import { waitForZtpfwPodToBeRecreated } from './utils';

const waitForClusterOperator = async (
  setError: (error: PersistErrorType) => void,
  name: string,
): Promise<boolean> => {
  console.info('waitForClusterOperator started for: ', name);
  for (let counter = 0; counter < MAX_LIVENESS_CHECK_COUNT; counter++) {
    try {
      console.log('Querying co: ', name);
      const operator = await getClusterOperator(name).promise;
      if (
        getCondition(operator, 'Progressing')?.status === 'False' &&
        getCondition(operator, 'Degraded')?.status === 'False' &&
        getCondition(operator, 'Available')?.status === 'True'
      ) {
        // all good
        setError(null);
        return true;
      }
    } catch (e) {
      console.error('waitForClusterOperator error: ', e);
      // do not report, keep trying

      // setError({
      //   title: WAIT_ON_OPERATOR_TITLE,
      //   message: `Failed to query status of ${name} cluster operator`,
      // });
    }

    await delay(DELAY_BEFORE_QUERY_RETRY);
  }

  setError({
    title: WAIT_ON_OPERATOR_TITLE,
    message: `Failed to query status of ${name} cluster operator on time.`,
  });

  return false;
};

const waitOnReconciliation = async (
  setError: (error: PersistErrorType) => void,
  setProgress: UsePersistProgressType['setProgress'],
  state: K8SStateContextData,
  persistIdpResult: PersistIdentityProviderResult,
): Promise<boolean> => {
  const { domain } = state;

  // wait till the UI pod is recreated after changing domain (Deployment + route)
  if (!(await waitForZtpfwPodToBeRecreated(MAX_LIVENESS_CHECK_COUNT, domain))) {
    setError({
      title: UI_POD_NOT_READY,
      message: 'The configuration pod did not become ready on time.',
    });
    return false;
  }
  setProgress(PersistSteps.ReconcileUIPod);

  if (persistIdpResult === PersistIdentityProviderResult.userCreated) {
    // wait for API
    if (!(await waitForClusterOperator(setError, 'openshift-apiserver'))) {
      return false;
    }
    setProgress(PersistSteps.ReconcileApiOperator);

    // wait for identity provider
    if (!(await waitForClusterOperator(setError, 'authentication'))) {
      return false;
    }
    setProgress(PersistSteps.ReconcileAuthOperator);

    // TODO: openshift console??
  }

  if (!(await waitForClusterOperator(setError, 'kube-apiserver'))) {
    return false;
  }

  // Important: keep following aligned with the last reconcile-step
  setProgress(PersistSteps.ReconcileAuthOperator);

  console.info('waitOnReconciliation finished successfully');
  return true;
};

export const persist = async (
  state: K8SStateContextData,
  setError: (error: PersistErrorType) => void,
  setProgress: UsePersistProgressType['setProgress'],
  onSuccess: () => void,
) => {
  const persistIdpResult = await persistIdentityProvider(
    setError,
    setProgress,
    state.username,
    state.password,
  );

  if (persistIdpResult === PersistIdentityProviderResult.error) {
    console.error('Failed to persist IDP, giving up.');
    return;
  }

  if (persistIdpResult === PersistIdentityProviderResult.userCreated) {
    // Let the operator reconciliation start
    await delay(DELAY_BEFORE_FINAL_REDIRECT);

    if (!(await waitForClusterOperator(setError, 'authentication'))) {
      return false;
    }
  }

  console.log('Saving of IDP is over, about to continue with Domain.');
  if (!(await persistDomain(setError, setProgress, state.domain, state.customCerts))) {
    return;
  }

  console.log('Domain persisted, blocking progress till reconciled.');
  // Let the operator reconciliation start
  await delay(DELAY_BEFORE_FINAL_REDIRECT);
  if (!(await waitForClusterOperator(setError, 'authentication'))) {
    return false;
  }

  if (
    (await saveIngress(setError, setProgress, state.ingressIp)) &&
    (await saveApi(setError, setProgress, state.apiaddr))
  ) {
    // finished with success
    console.log('Data persisted, blocking progress till reconciled');

    setError(null); // show the green circle of success

    if (!(await waitOnReconciliation(setError, setProgress, state, persistIdpResult))) {
      return;
    }

    // delete route backup
    // TODO

    onSuccess();
  }
};

export const navigateToNewDomain = async (domain: string, contextPath: string) => {
  // We can not check livenessProbe on the new domain due to CORS
  // We can not use pod serving old domain either since it will be terminated and the route changed
  // So just wait...
  const ztpfwUrl = `https://${ZTPFW_UI_ROUTE_PREFIX}.apps.${domain}${contextPath}`;
  console.info('Changes are persisted, about to navigate to the new domain: ', ztpfwUrl);
  // We should go with following:
  window.location.replace(ztpfwUrl);
};
