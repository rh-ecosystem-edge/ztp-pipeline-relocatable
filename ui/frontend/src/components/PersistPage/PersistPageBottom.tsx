import React from 'react';
import { useNavigate } from 'react-router-dom';
import {
  Alert,
  AlertVariant,
  Button,
  ButtonVariant,
  Stack,
  StackItem,
} from '@patternfly/react-core';
import { CheckCircleIcon, InProgressIcon } from '@patternfly/react-icons';
import {
  global_primary_color_light_100 as progressColor,
  global_success_color_100 as successColor,
} from '@patternfly/react-tokens';

import { navigateToNewDomain, persist } from './persist';
import { PersistErrorType } from './types';
import { useK8SStateContext } from '../K8SStateContext';

import './PersistPageBottom.css';

export const PersistPageBottom: React.FC = () => {
  const navigate = useNavigate();
  const [error, setError] = React.useState<PersistErrorType>(/* undefined */);
  const [retry, setRetry] = React.useState(true);
  const state = useK8SStateContext();

  React.useEffect(() => {
    if (!retry) {
      return;
    }
    setRetry(false);
    setError(undefined);

    persist(state, setError, () => navigateToNewDomain(state.domain, '/wizard/final'));
  }, [retry, setError, state]);

  return (
    <Stack className="persist-page-bottom" hasGutter>
      {error?.title ? (
        <>
          <StackItem isFilled className="persist-page-bottom__error-item">
            <Alert variant={AlertVariant.danger} title={error.title} isInline>
              {error.message}
            </Alert>
          </StackItem>
          <StackItem className="persist-page-bottom__error-item-footer">
            <div className="wizard-footer">
              <Button variant={ButtonVariant.primary} onClick={() => setRetry(true)}>
                Try again
              </Button>
              <Button variant={ButtonVariant.link} onClick={() => navigate(`/wizard/sshkey`)}>
                Go back
              </Button>
            </div>
          </StackItem>
        </>
      ) : error === undefined ? (
        <>
          <StackItem>
            <InProgressIcon
              color={progressColor.value}
              className="persist-page-bottom__progress-icon"
            />
          </StackItem>
          <StackItem isFilled className="wizard-sublabel">
            Saving settings for your edge cluster...
          </StackItem>
        </>
      ) : (
        <>
          <StackItem>
            <CheckCircleIcon
              color={successColor.value}
              className="persist-page-bottom__success-icon"
            />
          </StackItem>
          <StackItem isFilled className="wizard-sublabel">
            Settings succesfully saved, waiting for them to take effect.
            {/* TODO: Show spinner */}
          </StackItem>
        </>
      )}
    </Stack>
  );
};
