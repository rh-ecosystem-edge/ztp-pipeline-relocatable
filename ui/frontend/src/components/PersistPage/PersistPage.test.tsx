import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { act } from 'react-dom/test-utils';

import { PersistPage } from './PersistPage';
import { K8SStateContextProvider } from '../K8SStateContext';
import { WizardProgressContextProvider } from '../WizardProgress';
import { delay } from '../../test-utils';

const Component: React.FC = () => (
  <K8SStateContextProvider>
    <WizardProgressContextProvider>
      <MemoryRouter>
        <PersistPage />
      </MemoryRouter>
    </WizardProgressContextProvider>
  </K8SStateContextProvider>
);

describe('PersistPage', () => {
  it('can render', async () => {
    let container;

    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(async () => {
      const { container: c } = render(<Component />);
      container = c;
      await delay(1000);
    });

    expect(container).toMatchSnapshot();
    // TODO: More complex scenario testing the persist() function on top of mocked data should be implemented
  });
});
