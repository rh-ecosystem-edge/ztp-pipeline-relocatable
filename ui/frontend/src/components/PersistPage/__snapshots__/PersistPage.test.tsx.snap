// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersistPage can render 1`] = `
<div>
  <div
    class="pf-c-background-image"
    style="--pf-c-background-image--BackgroundImage: url(/background.png); --pf-c-background-image--BackgroundImage-2x: url(/background.png); --pf-c-background-image--BackgroundImage--sm: url(/background.png); --pf-c-background-image--BackgroundImage--sm-2x: url(/background.png); --pf-c-background-image--BackgroundImage--lg: url(/background.png); --pf-c-background-image--Filter: url(#patternfly-background-image-filter-overlay0);"
  >
    <svg
      class="pf-c-background-image__filter"
      height="0"
      width="0"
      xmlns="http://www.w3.org/2000/svg"
    >
      <filter
        id="patternfly-background-image-filter-overlay0"
      >
        <fecolormatrix
          type="matrix"
          values="1 0 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 1 0"
        />
        <fecomponenttransfer
          color-interpolation-filters="sRGB"
          result="duotone"
        >
          <fefuncr
            tableValues="0.086274509803922 0.43921568627451"
            type="table"
          />
          <fefuncg
            tableValues="0.086274509803922 0.43921568627451"
            type="table"
          />
          <fefuncb
            tableValues="0.086274509803922 0.43921568627451"
            type="table"
          />
          <fefunca
            tableValues="0 1"
            type="table"
          />
        </fecomponenttransfer>
      </filter>
    </svg>
  </div>
  <header
    class="pf-c-page__header page-header"
  >
    <div
      class="pf-c-page__header-brand"
    >
      <a
        class="pf-c-page__header-brand-link"
      >
        <img
          alt="Logo"
          src="RedHatLogo.svg"
        />
      </a>
    </div>
    <div
      class="pf-c-page__header-nav"
    >
      <div
        class="page-header-middle"
      >
        Edge cluster setup
      </div>
    </div>
    <div
      class="pf-c-page__header-tools"
    >
      <a
        aria-disabled="false"
        class="pf-c-button pf-m-tertiary"
        data-ouia-component-id="OUIA-Generated-Button-tertiary-1"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe="true"
      >
        Log out
      </a>
    </div>
  </header>
  <div
    class="page-container"
  >
    <div
      class="page-content"
    >
      <div
        class="pf-l-stack pf-m-gutter"
      >
        <div
          class="pf-l-stack__item content-two-rows__top"
        >
          <div
            class="welcome-top"
            style="background-image: url(/backgroundWelcome.png);"
          >
            <img
              alt="Kube"
              class="welcome-top__kube"
              src="kube.svg"
            />
          </div>
        </div>
        <div
          class="pf-l-stack__item content-two-rows__bottom"
        >
          <div
            class="pf-l-stack pf-m-gutter persist-page-bottom"
          >
            <div
              class="pf-l-stack__item"
            >
              <svg
                aria-hidden="true"
                class="persist-page-bottom__progress-icon"
                fill="#73bcf7"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 1024 1024"
                width="1em"
              >
                <path
                  d="M513.417211,16.013668 L513.417211,112.013668 C513.406007,120.539653 520.082422,127.576718 528.597211,128.013668 C732.697211,136.513668 896.147211,305.013668 896.147211,511.343668 C896.147211,723.013668 724.007211,895.163668 512.417211,895.163668 C437.186239,895.277345 363.602574,873.135095 300.927211,831.523668 C294.58293,827.2982 286.138663,828.135095 280.747211,833.523668 L211.807211,902.353668 C208.502935,905.676185 206.82123,910.280199 207.20607,914.950237 C207.590911,919.620275 210.00361,923.886884 213.807211,926.623668 C300.709573,989.398037 405.213535,1023.13146 512.417211,1023.01398 C794.537211,1023.01398 1023.91724,793.433668 1023.91724,511.413668 C1024.00721,235.103668 804.007211,9.22366802 529.897211,0.00645736761 C525.571205,-0.116171039 521.377128,1.51204372 518.271915,4.52681369 C515.166702,7.54158367 513.415263,11.6857144 513.417211,16.013668 M97.1272107,212.923668 C64.2190689,258.479443 39.0698895,309.162997 22.6972107,362.923668 C21.2705092,367.756559 22.2039184,372.979173 25.2163356,377.018685 C28.2287528,381.058197 32.9681443,383.442545 38.0072107,383.453683 L139.407211,383.453683 C145.970195,383.46249 151.873012,379.462528 154.297211,373.363668 C164.307669,347.593304 177.068112,322.977652 192.357211,299.943668 C196.579666,293.633658 195.741339,285.216847 190.357211,279.863668 L121.497211,211.013668 C118.186635,207.662531 113.567544,205.940538 108.871166,206.30669 C104.174788,206.672841 99.8783994,209.089927 97.1272107,212.913668 M112.907211,511.433663 L17.0072107,511.433663 C12.679257,511.433663 8.5351263,513.183159 5.52035633,516.288373 C2.50558635,519.393586 0.877371594,523.587662 1.00721065,527.913668 C4.13104542,629.408217 37.6085847,727.635936 97.1172107,809.913668 C99.8539943,813.717268 104.120603,816.129968 108.790642,816.514809 C113.46068,816.899649 118.064693,815.217944 121.387211,811.913668 L190.217211,743.073668 C195.605784,737.682216 196.442679,729.237949 192.217211,722.893668 C153.520825,664.580171 131.611534,596.745437 128.887211,526.813668 C128.553964,518.220138 121.487197,511.427209 112.887211,511.433663"
                />
              </svg>
            </div>
            <div
              class="pf-l-stack__item pf-m-fill wizard-sublabel"
            >
              Saving settings for your edge cluster...
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
